File: service-security-message/api/src/main/java/org/delivery/api/common/exception/ApiException.java
package org.delivery.api.common.exception;

import lombok.Getter;
import org.delivery.api.common.error.ErrorCodeIfs;


// 생성자 오버로딩, 여러 방식으로 ApiException 인스턴스 생성
@Getter
public class ApiException extends RuntimeException implements ApiExceptionIfs {

    private final ErrorCodeIfs errorCodeIfs;
    private final String errorDescription;

    public ApiException(ErrorCodeIfs errorCodeIfs){
        super(errorCodeIfs.getDescription());
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, String errorDescription){
        super(errorDescription);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorDescription;
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable tx){
        super(tx);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable tx, String errorDescription){
        super(tx);
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorDescription;
    }
}

File: service-security-kotlin/common/src/main/kotlin/org/delivery/common/exception/ApiException.kt
package org.delivery.common.exception

import org.delivery.common.error.ErrorCodeIfs

class ApiException : RuntimeException, ApiExceptionIfs {

    override val errorCodeIfs: ErrorCodeIfs

    override val errorDescription: String

    constructor(errorCodeIfs: ErrorCodeIfs) : super(errorCodeIfs.getDescription()) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(
        errorCodeIfs: ErrorCodeIfs, errorDescription: String
    ) : super(errorDescription) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorDescription
    }

    constructor(
        errorCodeIfs: ErrorCodeIfs, tx: Throwable
    ) : super(tx) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(errorCodeIfs: ErrorCodeIfs, tx: Throwable, errorDescription: String) : super(tx) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorDescription
    }
    
}